quinn/src/lib.rs

```diff
@@ -670,41 +453,85 @@ impl Pending {
 /// Stream of incoming connections.
 pub type Incoming = mpsc::Receiver<NewConnection>;
 
-/// A connection initiated by a remote client.
+/// A new connection (initiated locally or remotely)
 pub struct NewConnection {
+    /// The driver for the connection; this must be spawned for the connection to make progress
+    pub driver: ConnectionDriver,
     /// The connection itself.
     pub connection: Connection,
     /// The stream of QUIC streams initiated by the client.
     pub incoming: IncomingStreams,
 }
 
 impl NewConnection {
-    fn new(endpoint: Rc<RefCell<EndpointInner>>, handle: quinn::ConnectionHandle) -> Self {
-        let conn = Rc::new(ConnectionInner {
-            endpoint,
-            handle,
-            side: Side::Server,
-        });
-        NewConnection {
+    fn new(conn: Rc<RefCell<ConnectionInner>>) -> Self {
+        Self {
+            driver: ConnectionDriver(conn.clone()),
             connection: Connection(conn.clone()),
             incoming: IncomingStreams(conn),
         }
     }
 }
 
-/// A connection initiated locally.
-pub struct NewClientConnection {
-    /// The connection itself.
-    pub connection: Connection,
-    /// The stream of QUIC streams initiated by the client.
-    pub incoming: IncomingStreams,
+/// Connecting future
+pub struct ConnectingFuture(Option<ConnectionDriver>);
+
+impl Future for ConnectingFuture {
+    type Item = NewConnection;
+    type Error = io::Error;
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let connected = match &mut self.0 {
+            Some(driver) => {
+                match driver.poll() {
+                    Ok(Async::Ready(())) => unreachable!("cannot close without completing"),
+                    Ok(Async::NotReady) => {}
+                    Err(e) => {
+                        return Err(io::Error::new(io::ErrorKind::Other, format!("{}", e)));
+                    }
+                }
+                (driver.0).borrow().connected
+            }
+            None => panic!("polled after yielding Ready"),
+        };
+        if connected {
+            let ConnectionDriver(conn) = self.0.take().unwrap();
+            conn.borrow_mut().driver.take();
+            Ok(Async::Ready(NewConnection::new(conn)))
+        } else {
+            Ok(Async::NotReady)
+        }
+    }
 }
 
-impl NewClientConnection {
-    fn new(conn: Rc<ConnectionInner>) -> Self {
-        Self {
-            connection: Connection(conn.clone()),
-            incoming: IncomingStreams(conn.clone()),
+/// A future that drives protocol logic for a connection.
+#[must_use]
+pub struct ConnectionDriver(Rc<RefCell<ConnectionInner>>);
+
+impl Future for ConnectionDriver {
+    type Item = ();
+    type Error = io::Error;
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let conn = &mut *self.0.borrow_mut();
```

Comment:
Before proceeding, we should check `conn.pending.error` and return the error if one is found.

Reply:
Done.